feat(rendering): incremental rendering

sudo-tee opened the pull request:
Introduces an incremental rendering that updates only the changed portion(s) of the buffer instead of re-rendering everything on every change. This reduces CPU usages for large sessions.

This PR also incorporate a functional test suite in `tests/manual` where you can record a run snapshot tests.

cameronr commented:
I did a fairly big refactor and introduced a new `RenderState` object that subsumed _part_cache,  _message_map, and _actions to better handle finding parts/messages and tracking their rendered position. I think I'm happy with how that code is coming along :)

Separately, I haven't had a chance to look at snapshots/undo/redo yet but I've noticed a few times where canceling a request didn't result in `job_count == 0`. I have some ideas on that so I'm going to take a look at that next.

cameronr commented:
Nice! I think we're getting pretty close!

There's a FIXME in output_window about `restore_points`. Can you take a look at that whenever you next have a chance?

cameronr commented:
I have the code to render file mentions using the ranges in a part (by rerendering the earlier user prompt part) but I think there's bug in opencode where it sends back a range starting at 0 sometimes. You can see it in the `diff.json` test:

https://github.com/sudo-tee/opencode.nvim/blob/18374c114e15703d5e33a18500a9494ade258611/tests/data/diff.json#L97-L99

That causes the highlight to be added to the beginning of the user prompt. Interestingly, the opencode client seems to render that data the same way. Here's a similar session rendered in both:

in opencode.nvim:

<img width="859" height="236" alt="Screenshot 2025-10-21 at 17 16 53" src="https://github.com/user-attachments/assets/4313a95d-ea5b-4cfb-b3e5-0465de8096fb" />

in opencode tui:

<img width="916" height="244" alt="Screenshot 2025-10-21 at 17 16 47" src="https://github.com/user-attachments/assets/8077366c-38b8-4260-9cb0-54518678d04e" />

I can check and make sure the range starts with and `@` so at least it's not highlighting incorrectly but then there will be highlights missing. I guess, in that case, I could fall back to searching the text `value` and highlighting... I'll try that.

As just an aside, I also noticed that the opencode tui client has an off by one when rendering the highlight (I had the same bug initially):
<img width="766" height="167" alt="Screenshot 2025-10-21 at 16 14 32" src="https://github.com/user-attachments/assets/0af573ac-309a-4619-9386-65ebdf69a19f" />

sudo-tee commented:
This one was a bug in  the plugin itself.. there was a off by one error..  that I just fixed today in one of my commits.. you probably have an outdated file with the error

cameronr commented:
The off by one is in the opencode tui (not opencode.nvim). I was just pointing out that the opencode tui code doesn't seem very robust in this area.

All of that said, the off by one isn't important (after all, it's not even in our code).

The real issue is that the file ranges are sometimes / often zero which cause erroneous highlighting, both in the code I've just written and in the opencode tui app.

Here's some test data (with only the relevant parts):

```json
  {
    "type": "message.part.updated",
    "properties": {
      "part": {
        "sessionID": "ses_5f6d8a858ffeahzN9rkzzAlFMu",
        "messageID": "msg_a0929114f001p91aGAr5mfFVfh",
        "id": "prt_a0929114f002IHnVoMagRTlmq2",
        "type": "text",
        "text": "this is a mention test @diff-test.txt and @a-empty.txt. you can just say 1"
      }
    }
  },
  {
    "type": "message.part.updated",
    "properties": {
      "part": {
        "url": "file:///Users/cam/tmp/a/diff-test.txt",
        "sessionID": "ses_5f6d8a858ffeahzN9rkzzAlFMu",
        "mime": "text/plain",
        "messageID": "msg_a0929114f001p91aGAr5mfFVfh",
        "id": "prt_a09291150006rGf7yI937pvN4l",
        "type": "file",
        "filename": "diff-test.txt",
        "source": {
          "path": "/Users/cam/tmp/a/diff-test.txt",
          "type": "file",
          "text": {
            "start": 0,
            "end": 14,
            "value": "@diff-test.txt"
          }
        }
      }
    }
  },
  {
    "type": "message.part.updated",
    "properties": {
      "part": {
        "url": "file:///Users/cam/tmp/a/a-empty.txt",
        "sessionID": "ses_5f6d8a858ffeahzN9rkzzAlFMu",
        "mime": "text/plain",
        "messageID": "msg_a0929114f001p91aGAr5mfFVfh",
        "id": "prt_a092911500032HS2DGrZwukRwQ",
        "type": "file",
        "filename": "a-empty.txt",
        "source": {
          "path": "/Users/cam/tmp/a/a-empty.txt",
          "type": "file",
          "text": {
            "start": 0,
            "end": 12,
            "value": "@a-empty.txt"
          }
        }
      }
    }
  },
```

in opencode.nvim:

<img width="859" height="236" alt="Screenshot 2025-10-21 at 17 16 53" src="https://github.com/user-attachments/assets/994b293f-4453-41ea-8270-4cc33bc31c4d" />

and in the opencode tui app:

<img width="916" height="244" alt="Screenshot 2025-10-21 at 17 16 47" src="https://github.com/user-attachments/assets/673e7ac9-23b1-45d8-8f2f-5803ff23f8be" />

and here's what it looks like after I implemented the fallback I mentioned above (if `source.text.start == 0` and there's not an `@` at that position in the text, then search for `value`):

<img width="864" height="221" alt="Screenshot 2025-10-21 at 18 18 37" src="https://github.com/user-attachments/assets/9218432d-0b9e-4567-a503-87fec2b5a806" />

cameronr commented:
Wanted to check on a few things:

1. I just pushed my mention highlighting change. Let me know if it doesn't work the way you want/expect
2. I updated extmarks in the formatter (and elsewhere) to be 0 indexed instead of 1 indexed to be consistent with the nvim api. It shouldn't have an impact but I just wanted to give you a heads up.
3. I'm wondering if we should make the `on_data_rendered` callback that's used for markdown rendering have a configurable debounce delay in case people want to tune it (e.g. markview seems to work better with a longer debounce). Maybe something like `on_data_rendered_debounce`?

cameronr commented:
@sudo-tee Just checked in my change that will restart the opencode server if the user cancels three times in a row. Let me know if it looks ok or if there are any changes that I should make.

There isn't much left on my todo list for incremental rendering. I'll keep testing it but I'm wondering if there's other stuff on your list or if we should start thinking about how to get it released?

sudo-tee commented:
@cameronr 

The fix looks great to me :). 

I don't have anything that I want to add at the moment.

The sooner we can merge this the better. 

As for the release. I will try to use it as much as possible in the next day or 2. 

There is also some conflicts with master, but I don't think they is any major issues.

Thanks again for the big effort :)

sudo-tee commented:
@cameronr 

I fixed a bug in the permissions prompt. When the focus changes to the editor it should display the global shortcuts and not the buffer local ones. I also added the keymaps for permission on the output buffer as well

sudo-tee commented:
@cameronr 

I was testing the PR for a while this morning and found 2 things.

When there is a lot of events the UI locks up. I ran some profiling and saw this

<img width="1703" height="1094" alt="image" src="https://github.com/user-attachments/assets/43c63475-f50a-4781-b98f-35c211c9b2e8" />

I think nvim is choking to render this many events. We might need a "scheduler" , I tried to put more vim.schedule bit it did not help.


I also saw a lot of markdown flashing with render-markdown. I think we might need to put a larger debounce for markdown rendering. 

I will play a bit more with it and try to address the issues I found.

cameronr commented:
Huh, i guess just some GH weirdness. I pulled the latest changes :)

sudo-tee commented:
I did a comparison to main and there you can really see the issue on the incremental rendering branch

Incremental rendering
https://github.com/user-attachments/assets/9a246fec-f234-4963-a061-f7660b4624ad

Main
https://github.com/user-attachments/assets/27f1d5d4-674f-413c-b310-1bb501561be6

You can see that the ui locks up when rendering..

cameronr commented:
I looked more close at `state.messages` and `state.append`. No one's actually subscribed to `state.messages` so all of the deepcopy work is throwaway work anyway.

I think the best fix is to not use `state.append` when adding messages. It does mean that the state notification system won't work for new messages, I think that's fine because:

1. EventManager already has new message notifications (but server only)
2. renderer is who cares the most about messages and it's already handling things internally, including notifying other modules if they need to update based on messages
3. in the future, we could add a separate new_message notification if need. in fact, `state.last_message` is kind of that already

I also don't think we need the `vim.schedule` in `EventManager:emit`.

We should also inspect other uses of `state.append/remove` to make sure they aren't sources of extra / not useful work.

sudo-tee commented:
Thanks for the details, I didn't know about the profiler options.

Here is the profile.  The Issue I see is really calling the set_lines

<img width="1027" height="531" alt="image" src="https://github.com/user-attachments/assets/8b75de38-aca1-42d7-9cc8-5f59d0a33016" />


As per the deepcopy,  for the old value, it's there because if you give a reference you will have the new value.

I see that this could be a problem, but in my profiling this is not really a hot path.

The hot path is the part rerendering as the part grows, overwrites the lines of the same part over and over again. I think we could try to only append new lines instead.

cameronr commented:
Hmm, I think that means there are several performance issues.

I think the `deepcopy` is definitely a problem because it copies the entire messages table, along with every part, every time a message is added. That will get slower and slower the more messages you have. My `huge_session` file has 1021 messages so as more and more messages are added, it gets quite slow.

Your profile results don't really make sense to me as `nvim_buf_set_lines` should be fast so there must be something I'm not understanding / something else going on.

To help debug, can you do the following:

1. Can you share the captured events from a session that shows this? It'll be much easier if we're working with the same data. If it's a session that already exists, you can temporarily modify `renderer._render_full_session_data` to add messages and parts to `event_manager.captured_events`:
```lua
function M._render_full_session_data(session_data)
  M.reset()

  -- disable auto-scroll, makes loading a full session much faster
  M._disable_auto_scroll = true

  local revert_index = nil

  local event_manager = state.event_manager

  for i, msg in ipairs(session_data) do
    -- output:add_lines(M.separator)
    -- state.current_message = msg

    if state.active_session.revert and state.active_session.revert.messageID == msg.info.id then
      revert_index = i
    end

    M.on_message_updated({ info = msg.info }, revert_index)

    table.insert(event_manager.captured_events, { type = 'message.updated', properties = { info = msg.info } })

    for _, part in ipairs(msg.parts or {}) do
      table.insert(event_manager.captured_events, { type = 'message.part.updated', properties = { part = part } })

      M.on_part_updated({ part = part }, revert_index)
    end
  end

  if revert_index then
    M._write_formatted_data(formatter._format_revert_message(state.messages, revert_index))
  end

  -- re-enable Auto-scroll
  M._disable_auto_scroll = false
  M._scroll_to_bottom()
end
```

and then you can save the events with `require('opencode.ui.debug_helper').save_captured_events('some-file.json')`

2. What happens if you comment out `vim.api.nvim_buf_set_lines(windows.output_buf, start_line, end_line, false, lines)
` from `output_window`? I know it won't update the buffer, but I'm curious if that changes the performance or nt.
3. With the replay file, do you see the same performance issue when using the replay tool?

sudo-tee commented:
If I comment the set_lines the ui does not lock up and the thinking animation is smooth


[perf.json](https://github.com/user-attachments/files/23133835/perf.json)

I included a perf file. 

When I run the file in the replay renderer everything looks smooth.


I will try to load the ReplayRender with my full config

sudo-tee commented:
With my full config I have the issue now also in replay.
With the same profile as you

<img width="1695" height="1364" alt="image" src="https://github.com/user-attachments/assets/43cd440f-e78f-4e61-a8f7-b8b9c4326530" />

cameronr commented:
Hmm, I suspect some interaction with your config as I have markdown rendering, syntax highlighting, etc with the same file without any performance issues:

<img width="2056" height="1286" alt="Screenshot 2025-10-24 at 12 58 29" src="https://github.com/user-attachments/assets/00701163-607c-4215-9d6f-215b7c719678" />

<img width="2056" height="1286" alt="Screenshot 2025-10-24 at 12 58 33" src="https://github.com/user-attachments/assets/67c9e554-ac16-48cc-93cd-4910d0921226" />

Is your nvim config shared anywhere?

cameronr commented:
Markdown rendering likely to be heavily dependent on system performance and I agree that if the markdown debounce delay is too short, it can be pretty jarring.

FWIW, with your config, `ReplayAll 5` on my machine  (apple m4) is smooth (although the markdown background does flicker a bit). This test file is probably also particularly sensitive to markdown performance impacts because it's all markdown.

So maybe we should just make the default debounce larger?

Lastly and not important at all, but if you wanted, you could add a dev block to your lazy.nvim config that specifies the directory where you do your local plugin development, e.g.:

```lua
  performance = {
  --- ...
  },
  dev = {
    path = '~/Dev/neovim-dev/',
    fallback = false,
  },
```

Then you can just set `dev = true,` in a plugin spec, and it'll use the files in that dir.

cameronr commented:
I'm going to try implementing a text delta update when replacing a part. I had tried it early on but it introduced a bunch of bugs and I wasn't sure a dedicated code path was worth it.

I looked at perf.json, tho, and and a single part occurs over 2000:

```
❯ rg prt_ perf2.json | sort | uniq -c | sort
   1         "id": "prt_a17b4dc4c002cnV3qMuzC78qN0",
   1         "id": "prt_a17b4dc4f001l4IPCU1KYr2agj",
   1         "id": "prt_a17b4dc4f002OaKYJovbdWlZKb",
   1         "id": "prt_a17b4dc50001ZN7SRfxFasNXmK",
   1         "id": "prt_a17b4f1620013BhJdS3rKb37ML",
   1         "id": "prt_a17b51fb5001OJbu79hmP1ZzOv",
2189         "id": "prt_a17b4f16b001dtQr5PNp0YCP8u",
```

So I think it's worth another look.

sudo-tee commented:
Unfortunately I still see some locking. I think it's more the amount of writes rather the the size. It's like the editor needs to breathe. Maybe we can batch the same part

cameronr commented:
Ok, i fixed a bug where it was sometimes resetting the whole part so now it's both setting dramatically fewer lines and even fewer slightly fewer calls.

Original:
```
Event 2204/2204: session.updated ses_5e84ca285ffeY3zxSfonM8rH4T lines_set: 316076 set_calls: 2193
```

Latest
```
Event 2204/2204: session.updated ses_5e84ca285ffeY3zxSfonM8rH4T lines_set: 3868 set_calls: 1933
```

Can you try the following and see if any of them make a difference

1. Just this latest version. I've also included the debugging print code in output_window and the replay runner.
2. What happens if you turn off autocmds during wrapping `nvim_buf_set_lines` in `output_windows.set_lines` with `eventignore`:
```lua
  local old_eventignore = vim.o.eventignore
  vim.o.eventignore = 'all'
  vim.api.nvim_set_option_value('modifiable', true, { buf = windows.output_buf })
  -- vim.notify(vim.inspect(lines))
  vim.api.nvim_buf_set_lines(windows.output_buf, start_line, end_line, false, lines)
  vim.api.nvim_set_option_value('modifiable', false, { buf = windows.output_buf })
  vim.o.eventignore = old_eventignore
```
3. Is there any difference if you comment out setting extmarks and make output buf modifiable so you don't have toggle modifiable on and then off each time?
4. Is there any difference if you disable rendering of the footer, topbar, and loading animation?

I'm trying to figure out what's different between loading the full session (which makes all of the same event calls and then all of the same buffer calls) and having the events coming in more slowly.

It may be that we need to batch up changes but I'd really like to know which limit we're hitting to make sure we fix it.

sudo-tee commented:
> I spent more time looking at this and I think it's related to treesitter. When able, can you try just commenting out the treesitter register line:
> 
> https://github.com/sudo-tee/opencode.nvim/blob/e93708de4e1e25fa326193a940cb1f0a45dcd257/lua/opencode/ui/ui.lua#L86

Disabling tressitter makes the buffer fly... like in the Replay mode without the full config. 

> If that works, can you comment that line back in but try turning off treesitter at the start of `renderer.on_part_updated` via `vim.treesitter.stop(state.windows.output_buf)` and then calling `vim.treesitter.start(state.windows.output_buf)` in trigger_on_data_rendered? That will debounce the async treesitter processing. The experience isn't great because the colors disappear but I just want to see if that actually makes the perf issue go away.

Disabling it on part_render, makes the same effect, the CPU never spike over 10%.

> I kind of wonder if the faster we make the event processing, the worse it is because that just queues up more work for treesitter and eventually we hit some limit. That would also make sense why doing `ReplayFullSession` or `ReplayAll 0` doesn't exhibit the problem because we do all of our work in a tight loop (without any vim.schedules/defer_fn) so the main loop doesn't have a chance to get treesitter looking at each of our tiny changes.

This was my intuition all along also.  There is to many changes at the same time so probably treesitter is scheduling some work and starts it... then another changes happens and it starts all over again. With the full session there is no time period available for treesitter to schedule some work.


> Assuming that's it, I don't think it's a good experience to turn off treesitter while we're waiting for more updates. And since I don't know of another way to control it, I think we'll have to do some kind of throttling. I have a first pass at the code but I'm still figuring out a good way to test it. Should have something tomorrow.

The experience of disabling it on each part render, is not good I agree, but it still better then having the ui locking up. (Maybe we can offer a config for people, if they have issues).

I played also a with throttling yestarday, in my tests it was really helping, but it caused way to many ui issues with text written at the wrong place or with the wrong extmarks, so I scraped the idea, because I didn't want to rewrite the whole rendered.

But this led me to an idea for a future refactoring.

Having a `Line` object with it's extmark instead of having to keep a seperate list in sync. This way it would be much easier to batch updates. But this is for the future.

Thanks again for your effort.

sudo-tee commented:
I just did some testing and I can say that it's now reeeally fast :)  

This is amazing.

I think with this I can safely merge. 

Let me know if you still need to look at things before.

I will do some last testing either tonight or tomorrow.

Amazing work

Thanks a lot for this.

cameronr commented:
I have a little bit of debug logging to remove. After that, we should just give a few days of bake time to make sure I didn't break anything / no other issues pop up. Will be great to get it out!

I'm soooo glad we had the unit tests and replay system. It wouldn't have been possible to improve things so quickly without them!

sudo-tee commented:
On a different note.

I am thinking of changing the plugin name and maybe soon having release numbers. GitHub supports renaming project so it should be fine

There is already another plugin named opencode.nvim.

I have a couple of names and wanted your input.

- openkode.nvim
- nvim-opencode
- neocoder.nvim
- opencoder.nvim
- opencode-ui.nvim
- codepal.nvim
- libre-code.nvim

I'm open for ideas ⁉️

cameronr commented:
Oh funny, I didn't even know about the other one. Native neovim UI and keymaps/navigation is such an improvement over a terminal integration, at least in my opinion... but I'm biased :)

On names, here are some that stand out to me:
- opencode.nvim (leave it as is and let the community sort it out)
- opencode-native.nvim (highlight the native aspect of it)
- neopencode.nvim (it's a little cumbersome but i think maintaining the connection to opencode is important)
- neo-opencode (another take on the idea above).

I do think keeping a direct connection to opencode is important from a "branding" perspective.

cameronr commented:
I've just been cleaning up some diagnostics. I did change the emmyrc to not include `$HOME/.local/share/nvim/lazy` as that can introduce some type conflicts if plugins define the same types (that was happening to me). I think it's probably the better choice but let me know if you think otherwise.
